onnxruntime/core/providers/cpu/signal/dft.cc

```diff
@@ -0,0 +1,577 @@
+
+  U* window_data = nullptr;
+  if (window) {
+    window_data = const_cast<U*>(reinterpret_cast<const U*>(window->DataRaw()));
+  }
+
+  auto angular_velocity = compute_angular_velocity<T>(dft_length, inverse);
+
+  for (size_t i = 0; i < dft_output_size; i++) {
+    std::complex<T>& out = *(Y_data + i * Y_stride);
+    out.real(0);
+    out.imag(0);
+
+    for (size_t j = 0; j < dft_length; j++) {  // vectorize over this loop
+      auto exponential = compute_exponential(i * j, angular_velocity);
+      auto window_element = window_data ? *(window_data + j) : 1;
+      auto x = (j < number_of_samples) ? *(X_data + j * X_stride) : 0;
+      auto element = x * window_element;
+      out += exponential * element;
+    }
+
+    if (inverse) {
+      out /= static_cast<T>(dft_length);
+    }
+  }
+
+  return Status::OK();
+}
+
+template <typename T, typename U>
+static Status discrete_fourier_transform(OpKernelContext* ctx, const Tensor* X, Tensor* Y, int64_t axis,
+                                         int64_t dft_length, const Tensor* window, bool is_onesided, bool inverse,
+                                         InlinedVector<std::complex<T>>& V,
+                                         InlinedVector<std::complex<T>>& temp_output) {
+  // Get shape
+  const auto& X_shape = X->Shape();
+  const auto& Y_shape = Y->Shape();
+
+  auto batch_and_signal_rank = X->Shape().NumDimensions();
+  auto total_dfts = static_cast<size_t>(X->Shape().Size() / X->Shape()[axis]);
+
+  auto is_input_real = X->Shape().NumDimensions() == 2 || X->Shape()[X->Shape().NumDimensions() - 1] == 1;
+  auto complex_input_factor = is_input_real ? 1 : 2;
+  if (X->Shape().NumDimensions() > 2) {
+    total_dfts /= X->Shape()[X->Shape().NumDimensions() - 1];
+    batch_and_signal_rank -= 1;
+  }
+
+  // Calculate x/y offsets/strides
+  for (size_t i = 0; i < total_dfts; i++) {
+    size_t X_offset = 0;
+    size_t X_stride = X_shape.SizeFromDimension(axis + 1) / complex_input_factor;
+    size_t cumulative_packed_stride = total_dfts;
+    size_t temp = i;
+    for (size_t r = 0; r < batch_and_signal_rank; r++) {
+      if (r == static_cast<size_t>(axis)) {
+        continue;
+      }
+      cumulative_packed_stride /= X_shape[r];
+      auto index = temp / cumulative_packed_stride;
+      temp -= (index * cumulative_packed_stride);
+      X_offset += index * X_shape.SizeFromDimension(r + 1) / complex_input_factor;
+    }
+
+    size_t Y_offset = 0;
+    size_t Y_stride = Y_shape.SizeFromDimension(axis + 1) / 2;
+    cumulative_packed_stride = total_dfts;
+    temp = i;
+    for (size_t r = 0; r < batch_and_signal_rank; r++) {
+      if (r == static_cast<size_t>(axis)) {
+        continue;
+      }
+      cumulative_packed_stride /= X_shape[r];
+      auto index = temp / cumulative_packed_stride;
+      temp -= (index * cumulative_packed_stride);
+      Y_offset += index * Y_shape.SizeFromDimension(r + 1) / 2;
+    }
+
+    if (is_power_of_2(dft_length)) {
```

Comment:
I'm wondering what torch or tensorflow do when dft_length is not a power of two. Did you benchmark against their implementation?

Reply:
FFT in Torch is only built with MKL support, I dont see a vanilla CPU implementation of it.
From taking a look now, It seems that TF may be using a FFT implementation in Eigen.
It look like Eigen does Cooley-Tukey for powers of 2, and a Bluestein FFT algorithm for non-powers of 2.
I have not benchmarked against Eigen. It may be worthwhile doing, but I would say this is probably out of scope...